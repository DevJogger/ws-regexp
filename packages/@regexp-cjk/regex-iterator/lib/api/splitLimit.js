"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.splitLimit = void 0;
const _each_1 = __importDefault(require("./_each"));
function greedySplit(str, separator, limit) {
    const result = str.split(separator, limit);
    if (result.length === limit) {
        let length = 0;
        if (typeof separator === 'string') {
            length = result.join(separator).length;
        }
        else {
            length = result.reduce((l, x, i) => {
                let separatorLength = 0;
                if (i + 1 < limit) {
                    separatorLength = str.slice(l).match(separator).shift().length;
                }
                return l + x.length + separatorLength;
            }, 0);
        }
        result[limit - 1] += str.slice(length);
    }
    return result;
}
function _pushNonEmpty(arr, s, allowEmpty) {
    if (s !== '' || allowEmpty === true) {
        arr.push(s);
    }
}
function splitLimit(input, separator, limit, options) {
    if (typeof limit === 'object' && limit !== null) {
        options = limit;
        limit = options.limit;
    }
    if (limit === 0 || limit < 0 || isNaN(limit)) {
        limit = void 0;
    }
    if (typeof separator === 'string') {
        return greedySplit(input, separator, limit);
    }
    options = options !== null && options !== void 0 ? options : {};
    let ret = [];
    let lastIndex = 0;
    let lastString = '';
    options.global = true;
    let { useFullMatched, allowEmpty, excludeSubMatched } = options;
    useFullMatched = !!useFullMatched;
    allowEmpty = !!allowEmpty;
    excludeSubMatched = !!excludeSubMatched;
    for (let row of _each_1.default(input, separator, options)) {
        let { match, re } = row;
        let s;
        s = input.slice(lastIndex, match.index);
        _pushNonEmpty(ret, lastString = s, allowEmpty);
        //console.log(lastIndex, re.lastIndex, match.index)
        if (ret.length === limit) {
            lastIndex = match.index;
            break;
        }
        lastIndex = re.lastIndex;
        if (excludeSubMatched !== true && match.length > 1) {
            if (useFullMatched === true) {
                s = match[0];
                _pushNonEmpty(ret, lastString = s, allowEmpty);
            }
            else {
                for (s of match.slice(1)) {
                    _pushNonEmpty(ret, lastString = s, allowEmpty);
                    if (ret.length === limit) {
                        break;
                    }
                }
            }
        }
        if (ret.length === limit) {
            break;
        }
    }
    if (limit > 0) {
        if (!options.excludeRest) {
            let s = input.slice(lastIndex);
            if (ret.length < limit) {
                _pushNonEmpty(ret, s, allowEmpty);
            }
            else {
                ret[ret.length - 1] += input.slice(lastIndex);
            }
        }
    }
    else {
        _pushNonEmpty(ret, input.slice(lastIndex), allowEmpty);
    }
    return ret;
}
exports.splitLimit = splitLimit;
exports.default = splitLimit;
//# sourceMappingURL=splitLimit.js.map