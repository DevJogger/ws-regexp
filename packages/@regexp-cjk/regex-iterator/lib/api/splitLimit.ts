import _each from './_each';
import { ICloneRegexpOptionsCustom, ICloneRegexpOptions } from '@regexp-cjk/clone-regexp/lib/types';

function greedySplit(str: string, separator: string | RegExp, limit?: number)
{
	const result = str.split(separator, limit);
	if (result.length === limit)
	{
		let length = 0;
		if (typeof separator === 'string')
		{
			length = result.join(separator).length;
		}
		else
		{
			length = result.reduce((l, x, i) =>
			{
				let separatorLength = 0;
				if (i + 1 < limit)
				{
					separatorLength = str.slice(l).match(separator).shift().length;
				}
				return l + x.length + separatorLength;
			}, 0);
		}
		result[limit - 1] += str.slice(length);
	}
	return result;
}

function _pushNonEmpty(arr: string[], s: string, allowEmpty?: boolean)
{
	if (s !== '' || allowEmpty === true)
	{
		arr.push(s)
	}
}

export interface ISplitLimitOptions extends ICloneRegexpOptionsCustom
{
	useFullMatched?: boolean,
	excludeRest?: boolean,
	excludeSubMatched?: boolean,
	allowEmpty?: boolean,
	limit?: number,
}

export function splitLimit(input: string, separator: string | RegExp, limit?: number | ISplitLimitOptions, options?: ISplitLimitOptions): string[]
{
	if (typeof limit === 'object' && limit !== null)
	{
		options = limit;
		limit = options.limit
	}

	if (limit === 0 || limit < 0 || isNaN(limit as any))
	{
		limit = void 0;
	}

	if (typeof separator === 'string')
	{
		return greedySplit(input, separator, limit as number)
	}

	options = options ?? {};

	let ret: string[] = []
	let lastIndex = 0;
	let lastString = '';

	(options as ICloneRegexpOptions).global = true;

	let { useFullMatched, allowEmpty, excludeSubMatched } = options;
	useFullMatched = !!useFullMatched;
	allowEmpty = !!allowEmpty;
	excludeSubMatched = !!excludeSubMatched;

	for (let row of _each(input, separator, options))
	{
		let { match, re } = row;
		let s: string;

		s = input.slice(lastIndex, match.index)

		_pushNonEmpty(ret, lastString = s, allowEmpty);

		//console.log(lastIndex, re.lastIndex, match.index)

		if (ret.length === limit)
		{
			lastIndex = match.index;
			break;
		}

		lastIndex = re.lastIndex;

		if (excludeSubMatched !== true && match.length > 1)
		{
			if (useFullMatched === true)
			{
				s = match[0];

				_pushNonEmpty(ret, lastString = s, allowEmpty);
			}
			else
			{
				for (s of match.slice(1))
				{
					_pushNonEmpty(ret, lastString = s, allowEmpty);

					if (ret.length === limit)
					{
						break;
					}
				}
			}

		}

		if (ret.length === limit)
		{
			break;
		}
	}

	if (limit > 0)
	{
		if (!options.excludeRest)
		{
			let s = input.slice(lastIndex);

			if (ret.length < limit)
			{
				_pushNonEmpty(ret, s, allowEmpty)
			}
			else
			{
				ret[ret.length-1] += input.slice(lastIndex);
			}
		}
	}
	else
	{
		_pushNonEmpty(ret, input.slice(lastIndex), allowEmpty);
	}

	return ret;
}

export default splitLimit
