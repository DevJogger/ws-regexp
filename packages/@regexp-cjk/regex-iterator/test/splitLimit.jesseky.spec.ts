import splitLimit, { ISplitLimitOptions } from '../lib/api/splitLimit';

const str = 'user-agent: Mozilla/5.0, OS: Mac, Arch:  amd64';
const str2 = 'apple - pear  -  melon - banana';
const str3 = 'apple *-* pear  .+.  melon ^-^ banana';

describe('RegExp separator: ', function ()
{
	it('Split with RegExp correctly', function ()
	{
		expect(splitLimit(str, /:\s+/, -1)).toStrictEqual(['user-agent', 'Mozilla/5.0, OS', 'Mac, Arch', 'amd64']);
		expect(splitLimit(str, /:\s+/, 0)).toStrictEqual(['user-agent', 'Mozilla/5.0, OS', 'Mac, Arch', 'amd64']);
		expect(splitLimit(str, /:\s+/, 1)).toStrictEqual(['user-agent: Mozilla/5.0, OS: Mac, Arch:  amd64']);
		expect(splitLimit(str, /:\s+/, 2)).toStrictEqual(['user-agent', 'Mozilla/5.0, OS: Mac, Arch:  amd64']);
		expect(splitLimit(str, /:\s+/, 3)).toStrictEqual(['user-agent', 'Mozilla/5.0, OS', 'Mac, Arch:  amd64']);
		expect(splitLimit(str, /:\s+/, 4)).toStrictEqual(['user-agent', 'Mozilla/5.0, OS', 'Mac, Arch', 'amd64']);
		expect(splitLimit(str, /:\s+/, 5)).toStrictEqual(['user-agent', 'Mozilla/5.0, OS', 'Mac, Arch', 'amd64']);
	});
});

describe('String separator: ', function ()
{
	it('Split with String correctly', function ()
	{
		expect(splitLimit(str, ': ', -1)).toStrictEqual(['user-agent', 'Mozilla/5.0, OS', 'Mac, Arch', ' amd64']);
		expect(splitLimit(str, ': ', 0)).toStrictEqual(['user-agent', 'Mozilla/5.0, OS', 'Mac, Arch', ' amd64']);
		expect(splitLimit(str, ': ', 1)).toStrictEqual(['user-agent: Mozilla/5.0, OS: Mac, Arch:  amd64']);
		expect(splitLimit(str, ': ', 2)).toStrictEqual(['user-agent', 'Mozilla/5.0, OS: Mac, Arch:  amd64']);
		expect(splitLimit(str, ': ', 3)).toStrictEqual(['user-agent', 'Mozilla/5.0, OS', 'Mac, Arch:  amd64']);
		expect(splitLimit(str, ': ', 4)).toStrictEqual(['user-agent', 'Mozilla/5.0, OS', 'Mac, Arch', ' amd64']);
		expect(splitLimit(str, ': ', 5)).toStrictEqual(['user-agent', 'Mozilla/5.0, OS', 'Mac, Arch', ' amd64']);
	});
});

describe('separator with undefined or null: ', function ()
{
	it('separator with undefined or null correctly', function ()
	{
		expect(splitLimit(str, /:\s+/)).toStrictEqual(['user-agent', 'Mozilla/5.0, OS', 'Mac, Arch', 'amd64']);
		expect(splitLimit(str, /:\s+/, null)).toStrictEqual(['user-agent', 'Mozilla/5.0, OS', 'Mac, Arch', 'amd64']);
	});
});

// test with groups
describe('RegExp separator with groups but drop submatches: ', function ()
{
	it('Split with RegExp groups but drop submatches correctly', function ()
	{
		let options: ISplitLimitOptions = {
			excludeSubMatched: true,
		}

		expect(splitLimit(str2, /\s*(-)\s*/, -1, options)).toStrictEqual(['apple', 'pear', 'melon', 'banana']);
		expect(splitLimit(str2, /\s*(-)\s*/, 0, options)).toStrictEqual(['apple', 'pear', 'melon', 'banana']);
		expect(splitLimit(str2, /\s*(-)\s*/, 1, options)).toStrictEqual(['apple - pear  -  melon - banana']);
		expect(splitLimit(str2, /\s*(-)\s*/, 2, options)).toStrictEqual(['apple', 'pear  -  melon - banana']);
		expect(splitLimit(str2, /\s*(-)\s*/, 3, options)).toStrictEqual(['apple', 'pear', 'melon - banana']);
		expect(splitLimit(str2, /\s*(-)\s*/, 4, options)).toStrictEqual(['apple', 'pear', 'melon', 'banana']);
		expect(splitLimit(str2, /\s*(-)\s*/, 5, options)).toStrictEqual(['apple', 'pear', 'melon', 'banana']);
	});
});

describe('RegExp separator with groups and keep submatches: ', function ()
{
	const input = str2;

	const options: ISplitLimitOptions = {
		//isKeepSubmatches: true
	};

	const regex = /\s*(-)\s*/;

	it('Split with RegExp groups and keep submatches correctly', function ()
	{

		expect(splitLimit(input, regex, -1, options))
			.toStrictEqual(['apple', '-', 'pear', '-', 'melon', '-', 'banana']);
		expect(splitLimit(input, regex, 0, options))
			.toStrictEqual(['apple', '-', 'pear', '-', 'melon', '-', 'banana']);
		expect(splitLimit(input, regex, 1, options)).toStrictEqual(['apple - pear  -  melon - banana']);

	});

	describe(`limit`, () => {
		for (let limit = 2; limit <= 5; limit++)
		{
			it(`limit:${limit}`, function ()
			{
				expect(splitLimit(str3, regex, limit, options))
					.toMatchSnapshot();
			});
		}
	})

	it.skip('Split with RegExp groups and keep submatches correctly', function ()
	{
		expect(splitLimit(input, regex, 2, options)).toStrictEqual(['apple', '-', 'pear  -  melon - banana']);
		expect(splitLimit(input, regex, 3, options)).toStrictEqual(['apple', '-', 'pear', '-', 'melon - banana']);
		expect(splitLimit(input, regex, 4, options))
			.toStrictEqual(['apple', '-', 'pear', '-', 'melon', '-', 'banana']);
		expect(splitLimit(input, regex, 5, options))
			.toStrictEqual(['apple', '-', 'pear', '-', 'melon', '-', 'banana']);
	});

});

describe('RegExp separator contains multiple groups and keep submatches: ', function ()
{
	const input = str3;

	const options: ISplitLimitOptions = {
		//isKeepSubmatches: true
	};

	const regex = /\s*([^-+])([-+])([^-+])\s*/;

	it('Split with RegExp contains multiple groups and keep submatches correctly', function ()
	{
		expect(splitLimit(input, regex, -1, options))
			.toStrictEqual(['apple', '*', '-', '*', 'pear', '.', '+', '.', 'melon', '^', '-', '^', 'banana']);
		expect(splitLimit(input, regex, 0, options))
			.toStrictEqual(['apple', '*', '-', '*', 'pear', '.', '+', '.', 'melon', '^', '-', '^', 'banana']);
		expect(splitLimit(input, regex, 1, options))
			.toStrictEqual(['apple *-* pear  .+.  melon ^-^ banana']);
	})

	describe(`limit`, () => {
		for (let limit = 2; limit <= 5; limit++)
		{
			it(`limit:${limit}`, function ()
			{
				expect(splitLimit(input, regex, limit, options))
					.toMatchSnapshot();
			});
		}
	})

	it.skip('Split with RegExp contains multiple groups and keep submatches correctly', function ()
	{
		expect(splitLimit(input, regex, 2, options))
			.toStrictEqual(['apple', '*', '-', '*', 'pear  .+.  melon ^-^ banana']);
		expect(splitLimit(input, regex, 3, options))
			.toStrictEqual(['apple', '*', '-', '*', 'pear', '.', '+', '.', 'melon ^-^ banana']);
		expect(splitLimit(input, regex, 4, options))
			.toStrictEqual(['apple', '*', '-', '*', 'pear', '.', '+', '.', 'melon', '^', '-', '^', 'banana']);
		expect(splitLimit(input, regex, 5, options))
			.toStrictEqual(['apple', '*', '-', '*', 'pear', '.', '+', '.', 'melon', '^', '-', '^', 'banana']);
	});
});

