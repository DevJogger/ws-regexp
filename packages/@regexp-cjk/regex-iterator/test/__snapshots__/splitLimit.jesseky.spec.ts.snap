// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegExp separator contains multiple groups and keep submatches:  limit limit:2 1`] = `
Array [
  "apple",
  "*pear  .+.  melon ^-^ banana",
]
`;

exports[`RegExp separator contains multiple groups and keep submatches:  limit limit:3 1`] = `
Array [
  "apple",
  "*",
  "-pear  .+.  melon ^-^ banana",
]
`;

exports[`RegExp separator contains multiple groups and keep submatches:  limit limit:4 1`] = `
Array [
  "apple",
  "*",
  "-",
  "*pear  .+.  melon ^-^ banana",
]
`;

exports[`RegExp separator contains multiple groups and keep submatches:  limit limit:5 1`] = `
Array [
  "apple",
  "*",
  "-",
  "*",
  "pear  .+.  melon ^-^ banana",
]
`;

exports[`RegExp separator with groups and keep submatches:  limit limit:2 1`] = `
Array [
  "apple *",
  "-* pear  .+.  melon ^-^ banana",
]
`;

exports[`RegExp separator with groups and keep submatches:  limit limit:3 1`] = `
Array [
  "apple *",
  "-",
  "* pear  .+.  melon ^-^ banana",
]
`;

exports[`RegExp separator with groups and keep submatches:  limit limit:4 1`] = `
Array [
  "apple *",
  "-",
  "* pear  .+.  melon ^",
  "-^ banana",
]
`;

exports[`RegExp separator with groups and keep submatches:  limit limit:5 1`] = `
Array [
  "apple *",
  "-",
  "* pear  .+.  melon ^",
  "-",
  "^ banana",
]
`;
